packages/labs/forms/src/form-associated.ts

```diff
@@ -0,0 +1,456 @@
+/**
+ * @license
+ * Copyright 2024 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {type ReactiveElement} from '@lit/reactive-element';
+
+// #region Types
+
+export interface FormValueOptions<T> {
+  converter?: FormValueConverter<T>;
+}
+
+interface FormValueConverter<T> {
+  toFormValue(v: T): FormValue;
+  fromFormValue(v: FormValue): T;
+}
+
+export type FormValue = string | File | FormData | null;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Constructor<T = object> = new (...args: any[]) => T;
+
+export interface FormAssociated extends ReactiveElement {
+  /**
+   * Implement this method to validate the element. Return a `ValidityResult`.
+   */
+  _getValidity?(): ValidityResult;
+
+  /**
+   * Internal method to validate the element. This method should be called by
+   * the element when it's state might have changed in a way that would affect
+   * its validity.
+   *
+   * Calls `_getValidity` and sets the validity flags on the element's
+   * internals, then returns the result of `internals.checkValidity()`.
+   *
+   * It's often a good idea to call this method in the constructor to set the
+   * initial validity state.
+   */
+  _validate(): boolean;
+}
+type FormAssociatedInterface = FormAssociated;
+
+export interface FormAssociatedConstructor {
+  role?: ElementInternals['role'];
+
+  formAssociated: true;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  new (...args: any[]): FormAssociated;
+}
+
+interface ValidityResult {
+  flags: Partial<Omit<ValidityState, 'valid'>>;
+  message?: string;
+  anchor?: HTMLElement;
```

Comment:
I believe this breaks in webkit, if anchor is omitted (or is the element itself, from what I remember).
https://github.com/whatwg/html/issues/10155
https://bugs.webkit.org/show_bug.cgi?id=269832

Reply:
This is apparently only a problem, if `setValidity` is called during constructor phase.